tools/networkmonitor/networkmonitor.nim

```diff
@@ -0,0 +1,230 @@
+
+    # store available user-agents in the network
+    if not allAgentStrings.hasKey(nodeUserAgent):
+      allAgentStrings[nodeUserAgent] = 0
+    allAgentStrings[nodeUserAgent] += 1
+
+    debug "connected to peer", peer=allPeers[customPeerInfo.peerId]
+   
+  # inform the total connections that we did in this round
+  let nOfOkConnections = allProtocols.len()
+  info "number of successful connections", amount=nOfOkConnections
+
+  # update count on each protocol
+  for protocol in allProtocols.keys():
+    let countOfProtocols = allProtocols[protocol]
+    peer_type_as_per_protocol.set(int64(countOfProtocols), labelValues = [protocol])
+    info "supported protocols in the network", protocol=protocol, count=countOfProtocols
+
+  # update count on each user-agent
+  for userAgent in allAgentStrings.keys():
+    let countOfUserAgent = allAgentStrings[userAgent]
+    peer_user_agents.set(int64(countOfUserAgent), labelValues = [userAgent])
+    info "user agents participating in the network", userAgent=userAgent, count=countOfUserAgent
+
+proc main() {.async.} = 
+  let conf: NetworkMonitorConf = NetworkMonitorConf.load()
+
+  info "cli flags", conf=conf
+
+  # TODO: Run sanity checks on cli flags, i.e. bootstrap nodes
+
+  if conf.logLevel != LogLevel.NONE:
+    setLogLevel(conf.logLevel)
+
+  if conf.metricsServer:
+    startMetricsServer(
+      conf.metricsServerAddress,
+      Port(conf.metricsServerPort))
+
+  let client = newHttpClient()
+
+  let
+    rng = keys.newRng()
+    bindIp = ValidIpAddress.init("0.0.0.0")
+    extIp = ValidIpAddress.init("127.0.0.1")
+    nodeKey1 = crypto.PrivateKey.random(Secp256k1, rng[])[]
+    nodeTcpPort1 = Port(60000)
+    nodeUdpPort1 = Port(9000)
+    node = WakuNode.new(
+      nodeKey1,
+      bindIp,
+      nodeTcpPort1)
+    
+    # TODO: Propose refactor removing initWakuFlags
+    flags = initWakuFlags(lightpush = false,
+                          filter = false,
+                          store = false,
+                          relay = true)
+
+  # TODO: use other discovery mechanisms: i.e. dns
+  
+  # mount discv5
+  node.wakuDiscv5 = WakuDiscoveryV5.new(
+      some(extIp), some(nodeTcpPort1), some(nodeUdpPort1),
+      bindIp,
+      nodeUdpPort1,
+      conf.bootstrapNodes,
+      false,
+      keys.PrivateKey(nodeKey1.skkey),
+      flags,
+      [],
+      node.rng
+    )
+
+  let d = node.wakuDiscv5.protocol
+  d.open()
+
+  # list of peers that we have discovered/connected
+  var allPeersRef = CustomPeersTableRef()
+
+  let serverAddress = initTAddress(conf.metricsRestAddress & ":" & $conf.metricsRestPort)
+  proc validate(pattern: string, value: string): int =
+    if pattern.startsWith("{") and pattern.endsWith("}"): 0
+    else: 1
+  var router = RestRouter.init(validate)
+  router.installHandler(allPeersRef)
+
+  # rest server for custom metrics
+  var sres = RestServerRef.new(router, serverAddress)
+  let restServer = sres.get()
+  restServer.start()
+
+  while true:
```

Comment:
sure, both comments are fair, will refactor that.

Regarding the use of timers, i have some concerns. imho using setTimer() to schedule a recurrent task as done in [here](https://github.com/status-im/nwaku/blob/master/waku/v2/node/waku_metrics.nim#L74-L76) seems convoluted to me and a bit hackish. I understand you are asking to do the same here. My reasons to think that:

* on each iteration you schedule again and again the task, x time from now, some kind of recursive setTimer.
* you need an external timer to trigger the function the first time and then a second internal timer that is called, again and again.
* this adds extra code and its more difficult to reason.
* in general the amount of code is more than simply `while true:` and `sleepAsync` with an `asyncSpawn`.
* seems hackish as we are constantly scheduling the task x time from now.
* i don't see nim-libp2p nor nimbus using `setTimer` for scheduling recurrent tasks.
* [nim-libpb pubsub loop](https://github.com/status-im/nim-libp2p/blob/fa5d102370cff4e68b47ac633153b22609b8aa41/libp2p/daemon/daemonapi.nim#L1286) follows the pattern i suggest. [same in this example](https://github.com/status-im/nim-libp2p/blob/fa5d102370cff4e68b47ac633153b22609b8aa41/examples/go-daemon/bootstrap.nim)
* same for [nimbus](https://github.com/status-im/nimbus-eth2/blob/97a1cdc4902eac0149b0e2f8a0ac9fa663afe408/beacon_chain/nimbus_beacon_node.nim#L1386)

As far as I know, we don't have a `ScheduleAtFixedRate(func, time)` like in other languages, so imho the best way would be to follow what nimlibp2p/nimbus does.

```nim
proc recurrentTask() {.async.} = 
  echo "do stuff"
  sleepAsync(5000)
```

and somewhere else 
```nim
asyncSpawn recurrentTask()
```

@LNSD thoughts?

Reply:
I'm afraid I disagree with your rationale about the timer tasks. But I will base my reasoning on facts instead of opinions. 

Short answer and a rule of thumb: 
> Using a `sleep` is **always** a hack unless it is strictly unavoidable. Some exceptions could be building a task scheduler, an OS, or working on bare metal system programming.

The fact that _nimbus_ and _nim-libp2p_ don't do it could be explained by other reasons we don't know. For example, amongst other things, `nim-libp2p` uses exception, and the status nim style guide [recommends the opposite](https://status-im.github.io/nim-style-guide/errors.exceptions.html#:~:text=In%20general%2C%20prefer%20explicit%20error%20handling%20mechanisms.).

So nothing done in other projects should be taken as written in stone.

--------------

### Some context

Checking the implementation of `asyncSleep()` it registers a timer in the `nim-chonos` scheduler:

https://github.com/status-im/nim-chronos/blob/6525f4ce1d1a7eba146e5f1a53f6f105077ae686/chronos/asyncloop.nim#L983-L1000

Also, as @rymnc pointed out in [this PR](https://github.com/status-im/nwaku/pull/1094#discussion_r957177210), there is a (stale) issue in `nim-chronos` that will bring (or not 🤷🏼) the desired interval timer primitive to the party like other technologies have (e.g., NodeJS) given that it is a common task: https://github.com/status-im/nim-chronos/pull/62

### Both approaches comparison

Let's compare the _while/sleep approach_ and the _interval/set timer_ approaches:

The **nested timers approach**, depending on how you implement it, can make a repetitive task start to be deterministically executed, i.e., if, at the moment of executing the callback, the first thing you do is to re-schedule the following execution, you can be sure that it will be executed next time at t+<interval you configured> (of course if the `nim-chronos`' dispatcher is not busy with any blocking workload), independently of the duration of that callback's task duration. This approach can also make the time separation between task executions deterministic by re-scheduling the task after the task is executed.

On the other hand, the **`while true` and sleep** approach: Requires to be within an async method that should be spawned by `asyncSpawn`. And the start of the execution is not deterministic. What is deterministic is the spacing between the different task executions; i.e., if your task takes X to be executed, then the next task will be executed at X+sleep time, and the following at X+sleep time+X'+sleep time (NB that we are working with IO so the X time can vary significantly).

So the first approach is objectively more versatile and covers scenarios that the second does not cover out of the box.

-------------

I hope this makes it clear why.

Writing this down takes a lot of effort. I know that it is helpful to understand the reasons for our suggestions, but it is time-consuming and hinders other team members' productivity. When someone suggests something, I start thinking, "why?" and look for the answers/facts to understand it.

I value your opinion, but we are doing engineering, which should be based on facts. And also, please, be mindful of others' time.

Reply:
Thanks for the heads up about minding others' time, but it's up to each one of us to allocate time wisely, so keep that in mind.

While I agree with your "fact-based arguments" I don't agree with your rule of thumb. Engineering is also pragmatic, so since i) I don't need deterministic time spacing and ii) `sleepAsync`/`asyncSpawn` is simpler and less convoluted than timers, that should be sufficient for this PR.

For sure I will consider using [interval](https://github.com/status-im/nim-chronos/pull/62) if merged and if I need a deterministic time spacing I will use your solutions with timers.

Thanks for the detailed explanation.